### DISCUSSION TITLE
App Gateway Diagram

### DISCUSSION THREAD

**Discussion by esdcharles** [Azure/ipam#108] [Q&A] [answered]
I am currently exploring the use of an App Gateway with the hope that this will address the need for IPAM to be a fully private solution and hopefully without raising the cost by too much. I am bit confused at this point because I am seeing 2 Web Apps in the diagram (IPAM UI and IPAM Engine) using a single IPAM App Service Plan. Are we supposed to have 2 Web Apps each running the same Service Plan (IPAM APP Service Plan?)  Before I move on to the SSL cert, routing rules (unclear at the moment but assuming the same as Front Door user) etc. I was referencing another user who decided to go with a Azure FD and they had mentioned creating separate RGs and that just confused me further lol Can you let me know if I am on the right path with my configuration thus far and if there is a preferred order?  Perhaps by labelling the diagram for others who wish to explore this route may be helpful.

Thanks for your continued assistance

Azure FD Post: 

https://github.com/Azure/ipam/discussions/90

Diagram I'm following:

![image](https://user-images.githubusercontent.com/102684496/229603607-7f64f72d-32a6-4784-b71b-e66a9921f588.png)

---

AppGW FrontEnd:

IP: Public IP

listener: Public IP

AppGW Backend:

Key Vault PE (VNet: IPAM NETWORK  RG: IPAMRG DNS: Private DNS Solution)

Cosmos PE (VNet: IPAM NETWORK  RG: IPAMRG DNS: Private DNS Solution)

IPAM WebAPP:

IPAM WebAPP (RG: IPAMRG) configured with IPAMAppServicePlan (RG: IPAMRG)

Private DNS: To Do

SSL Cert : To Do

VNet Integration: To do last?

Routing Rules: Paths to the UI /admin /discover /analyze /* (taken from other post)

---

**Comment by DCMattyG** [comment, marked as answer]
Hello @binarybfc, a lot to break down here, but I'll do my best.....

That diagram is provided as a **reference**. There are many ways to make the Azure IPAM a private solution and it's impossible to document every possible permutation. I created that diagram to demonstrate at a high-level a way that should work well for most folks, but of course there are always little differences here and there between customer environments.

First off, yes you can run both Apps in a single App Service Plan. The default deployment leverages a multi-container App Service, which does **not** support vNET integration. By splitting this into 2x App Services (one container running in each), they can be vNET integrated...however this now introduces the requirement for path-based routing, which is where App Gateway comes in.

High-level order is as follows....

1. Create networking resources (vNET/Subnets)
2. Create Private DNS Zones for all required resources (App Services, Blob Storage, Cosmos DB, KeyVault, Azure Container Registry)
3. Create Azure Resources (App Service Plan, 2x App Services, KeyVault, Cosmos DB, Container Registry, App Gateway V1/V2)
4. Create Private Endpoints for all resources and ties them to the matching Private DNS Zones
5. Configure KeyVault with required Secrets
6. Configure ACR and push container images
7. Configure App Services with KeyVault references in Application Settings
8. Configure App Services to pull containers from Azure Container Registry
9. Configure App Gateway w/ SSL certs and path-based routing ("/" goes to UI, and "/api" goes to Engine)

There more to it at a granular level, but that's the basics of what needs to be done.

Of course I'm happy to help you get more specific, and we will be better documenting this in the next 30 days or so. I'm sure I missed a few things above, so please feel free to let me know what other items need additional clarity and I'm more than happy to share my thoughts.

**Reply by esdcharles** [reply]
Thanks so much! I got everything completed today and am marking up the steps provided above. They were extremely helpful - especially for a guy who's Web App experience is slim to none. 

Regards,
@binarybfc

**Reply by DCMattyG** [reply]
Awesome @binarybfc! So wonderful to hear. I hope you enjoy everything, but please let us know if there anything else we can do for you!

**Reply by omerzubair** [reply]
Hi @binarybfc , is it possible to share the IaC solution for your complete solution. It will be great assistance. cheers

**Reply by omerzubair** [reply]
Hi @DCMattyG / @binarybfc ,
I just understood this line after making everything private and still was getting error (see screenshot below)

_"First off, yes you can run both Apps in a single App Service Plan. The default deployment leverages a multi-container App Service, which does not support vNET integration."_

![image](https://github.com/Azure/ipam/assets/7728453/798d6dde-fb2c-4d7f-8a56-1e6bf55d9416)

I get this same error when I use kudu/scm site. I guess it would be awesome if @DCMattyG can help with steps written above :) esp on how can path based rules and also setup the 2x web apps instead of one?

Cheers

**Reply by DCMattyG** [reply]
Hey @omerzubair, I'm more than happy to spend some 1:1 time with you looking into this. That said, I'm about to release v3 which has been compressed down to a single container so no more multi-container required, which also means Private Endpoints and vNET Integration will be fully supported as well without the need for an App Gateway in front.

The code push for v3 should be done by the end of this week.

**Reply by omerzubair** [reply]
this is the best news for my team here downunder :) Happy to be the test bed and revert with findings. 
Thanks for the 1:1 offer will explore it if I get stuck.

**Reply by omerzubair** [reply]
Hey Bud , 
Just reachingout for any update you might have ? You think we might have something sometime COB? Cheers

**Reply by DCMattyG** [reply]
Hey @omerzubair, apologies for the delayed response. Today is a holiday in the US.

I'm just wrapping up the documentation revisions. I'm happy to provide you early access to the new version if you'd like to reach out me via email at Matthew.Garrett@microsoft.com. I can provide you the details as to how to test the new branch and perhaps you can provide me some feedback if there are any changes which need to be made.

**Reply by NikolaiKleppe** [reply]
Hi @DCMattyG , I'm are interested in this update as well. I cant find any releases on the github page for v3 or other versions, is this update still under development? Which branch is it?

> without the need for an App Gateway in front

I assume it's still an option to put App Gateway in front if you want it reachable from external sources

**Reply by DCMattyG** [reply]
Hi @NikolaiKleppe, the code has been merged into main, however I'm still working through some GitHub action permission issues that will publish the Releases.

Yes, you can absolutely put an AppGateway in front still, but no more need for path-based routing as there will just be a single container with the new code. Before, the ONLY way to make this private was to split the containers, add Private Endpoints & Virtual Network Integration, then put AppGateway in the front for path-based routing. Now you can simply add a Private Endpoint & Virtual Network integration as there is a single container and the code does all of the routing internally.

Hope that helps!
